xls/modules/rle/rle.x

```diff
@@ -0,0 +1,306 @@
+pub struct EncOutData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // compressed symbol
+    count: bits[COUNT_WIDTH],   // symbol count
+    last: bool,                 // if this is the last symbol
+}
+
+// structure to preserve the state of an RLE encoder
+pub struct RLEEncState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // symbol from the previous RLEEnc::next evaluation,
+    // valid if prev_count > 0
+    prev_symbol: bits[SYMBOL_WIDTH],
+    // symbol count from the previous RLEEnc::next evaluation.
+    // zero means that the previous evaluation sent all the data and
+    // we start counting from the beginning
+    prev_count: bits[COUNT_WIDTH],
+    // flag indicating that the previous symbol was the last one
+    // in the transmission
+    prev_last: bool,
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<EncInData<SYMBOL_WIDTH>> in;
+    output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev_symbol: bits[SYMBOL_WIDTH]:0,
+            prev_count: bits[COUNT_WIDTH]:0,
+            prev_last: false,
+        }
+    )}
+
+    config (
+        input_r: chan<EncInData<SYMBOL_WIDTH>> in,
+        output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = EncInData { symbol: bits[SYMBOL_WIDTH]:0, last: false };
+        let (input_tok, input) = recv_if(tok, input_r, !state.prev_last, zero_input);
+
+        let prev_symbol_valid = state.prev_count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (input.symbol != state.prev_symbol);
+        let overflow = state.prev_count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let (symbol, count, last) = if (state.prev_last) {
+            (
+                bits[SYMBOL_WIDTH]:0,
+                bits[COUNT_WIDTH]:0,
+                false
+            )
+        } else if (symbol_differ || overflow) {
+            (
+                input.symbol,
+                bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        } else {
+            (
+                input.symbol,
+                state.prev_count + bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        };
+
+        let do_send = state.prev_last || symbol_differ || overflow;
+        if do_send {
+            let data = EncOutData {
+                symbol: state.prev_symbol,
+                count: state.prev_count,
+                last: state.prev_last
+            };
+            let data_tok = send(input_tok, output_s, data);
+        } else {()};
+
+        RLEEncState {
+            prev_symbol: symbol,
+            prev_count: count,
+            prev_last: last,
+        }
+    }
+}
+
+// RLE encoder specialization for the codegen
+proc RLEEnc32 {
```

Comment:
is the specialization necessary? or can we just use `type` definition on procs? (if not, maybe we should log an issue to allow it?)

Reply:
A similar approach is used in the `delay.x` example:
https://github.com/google/xls/blob/main/xls/examples/delay.x#L202

The purpose of this specialization is to be able to generate the Verilog out of it, because currently it seems to be impossible to generate Verilog out of the parametric proc:
```
3 ir_converter_main.cc:163] Check failed: ::absl::OkStatus() == (status) (OK vs. NOT_FOUND: Top-level type info not found for proc "RLEEnc".)
```

Reply:
Oh I see, this does sound different from #959, it would be nice to have a way to alias procs like we do it with other type, see https://github.com/google/xls/blob/main/xls/examples/delay.x#L202.

Filed #998